[rebase of #845] rpc: add new btcd extension commands: addminingaddr, delminingaddr, listminingaddrs

yashbhutwala opened the pull request:
This is a rebase of #845 on master, closes #1411 😃

dannypaz commented:
Would love to have this functionality to test apps w/ btcd (neutrino). @Roasbeef

dannypaz commented:
FYI was seeing a `null pointer reference` when running this code in test. `s.server` returns null. The fix is to add a second arg `s server` to `newRpcServer` to set the `server` property for those methods.

jakesyl commented:
@jcvernaleo (as per #1530)

- Low priority
- Enhancement

dannypaz commented:
What needs to be done for this to be merged?

jcvernaleo commented:
@dannypaz if you could rebase and squash it that would definitely help on my end.  I know you rebased this a long time ago but since I'm trying to finally get some of these old PRs in there is going to be some churn.  This one is pretty well reviewed already so it is high on my list.

Rjected commented on rpcserver.go:
```diff
@@ -127,12 +127,13 @@ type commandHandler func(*rpcServer, interface{}, <-chan struct{}) (interface{},
 	"debuglevel":            handleDebugLevel,
 	"decoderawtransaction":  handleDecodeRawTransaction,
 	"decodescript":          handleDecodeScript,
-	"estimatefee":           handleEstimateFee,
```
Why is `estimatefee` removed here?

Rjected commented on server.go:
```diff
@@ -2549,6 +2551,117 @@ func setupRPCListeners() ([]net.Listener, error) {
+// miningAddrStore encapsulates a set of coinbase generation addresses in a
+// concurrent safe interface. All addresses stored MUST be unique, any
+// duplicates are rejected.
+type miningAddrStore struct {
+	miningAddrs []btcutil.Address
+	sync.RWMutex
+}
```
w.r.t my comment earlier about using an interface instead, `miningAddrStore` has little to do with the rest of what is happening in `server.go` and would be better suited in the `mining/cpuminer` package.

Rjected commented on rpcserver.go:
```diff
@@ -359,6 +361,42 @@ func handleAskWallet(s *rpcServer, cmd interface{}, closeChan <-chan struct{}) (
+	if err != nil {
+		return nil, &btcjson.RPCError{
+			Code:    btcjson.ErrRPCInvalidAddressOrKey,
+			Message: "Invalid address or key: " + err.Error(),
```
Should use a `fmt.Sprintf` here

Rjected commented on rpcserver.go:
```diff
@@ -359,6 +361,42 @@ func handleAskWallet(s *rpcServer, cmd interface{}, closeChan <-chan struct{}) (
+	if !addr.IsForNet(activeNetParams.Params) {
+		return nil, &btcjson.RPCError{
+			Code: btcjson.ErrRPCInvalidAddressOrKey,
+			Message: fmt.Sprintf("Mining address %v is on the "+
+				"wrong network: ", activeNetParams.Params),
```
Wouldn't this just substitute the `%v` with `activeNetParams.Params`? Maybe the line should be `fmt.Sprintf("Mining address %s is on the wrong network: %s", addr.String(), activeNetParams.Params.Name)` or something similar.

Rjected commented on rpcserver.go:
```diff
@@ -876,11 +914,47 @@ func handleEstimateFee(s *rpcServer, cmd interface{}, closeChan <-chan struct{})
+	if err := s.server.miningAddrs.RemoveAddr(addr); err != nil {
+		return nil, &btcjson.RPCError{
+			Code:    btcjson.ErrRPCInvalidAddressOrKey,
+			Message: fmt.Sprintf("Mining address %v not found", addr),
```
Would be useful to use the `String()` method on address here rather than just `%v`'ing it

Rjected commented on rpcserver.go:
```diff
@@ -876,11 +914,47 @@ func handleEstimateFee(s *rpcServer, cmd interface{}, closeChan <-chan struct{})
+	if err != nil {
+		return nil, &btcjson.RPCError{
+			Code:    btcjson.ErrRPCInvalidAddressOrKey,
+			Message: "Invalid address or key: " + err.Error(),
+		}
+	}
+
+	// Ensure the the passed address belongs to the same network that btcd
+	// is currently on.
+	if !addr.IsForNet(activeNetParams.Params) {
+		return nil, &btcjson.RPCError{
+			Code: btcjson.ErrRPCInvalidAddressOrKey,
+			Message: fmt.Sprintf("Mining address %v is on the "+
+				"wrong network: ", activeNetParams.Params),
```
Same comments as above, use `fmt.Sprintf` for line 926 and modify the `Message` to be more informative in line 935 and 936

Rjected commented on rpcserverhelp.go:
```diff
@@ -29,6 +29,10 @@ var helpDescsEnUS = map[string]string{
 	"debuglevel--result1":    "The list of subsystems",
 
+	// AddMiningAddrCmd help.
+	"addminingaddr--synopsis": "Adds an additional available mining address to the set of mining addresses for the daemon",
```
Maybe remove "for the daemon"

jakesyl commented:
Quick question on documenting these. As per #1546 went to add these to the [wiki](https://github.com/btcsuite/btcd/wiki/Bitcoind-RPC-Parity) and didn't see them in [bitcoin core](https://bitcoincore.org/en/doc/0.19.0/rpc/blockchain/getmempoolentry/). Anywhere I can get these add ons from to make sure we have parity here, or not relevant

jakesyl commented:
@yashbhutwala we're going to have to rebase against master to fix these conflicts and get these tests to stop running on travis

yashbhutwala commented:
@Rjected @jakesylvestre @dannypaz @jcvernaleo Thank you the review and activity here.  Sorry I haven't followed-up here in a long time.  I have rebased off of the latest master and updated the approach here to address @Rjected's previous review concerns.  Please re-review this PR, would love to get it merged 🙏

jcvernaleo reviewed (approved):
Looks good.

OK
